File: electron/engine/buttonIndex.ts
// Stable joystick-button index assignment — ported from server/src/codegen/button_index.rs.

import type { Control, Project } from "./types";

/** Lexicographic compare matching Rust's `str::cmp` (by code unit, not locale). */
function strCmp(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/** Number of logical joystick buttons a control contributes. */
function buttonCount(control: Control): number {
  switch (control.kind) {
    case "button":
      return 1;
    case "switch":
      return 2; // ON + OFF
    case "selector":
      return control.positions.length;
    case "encoder":
      return control.encoder.mode === "buttons" ? 2 : 0; // CW + CCW
    case "analog":
      return 0;
  }
}

/**
 * Map of `controlId` → (subIndex → 0-based joystick button number). Stable:
 * controls are ordered by label then id so indices survive regeneration.
 */
export function assignButtonIndices(
  project: Project,
  boardId: string,
): Map<string, Map<number, number>> {
  const controls = project.controls
    .filter((c) => c.boardId === boardId)
    .sort((a, b) => strCmp(a.label, b.label) || strCmp(a.id, b.id));

  const map = new Map<string, Map<number, number>>();
  let next = 0;
  for (const control of controls) {
    const count = buttonCount(control);
    const subMap = new Map<number, number>();
    for (let sub = 0; sub < count; sub++) {
      subMap.set(sub, next);
      next += 1;
    }
    map.set(control.id, subMap);
  }
  return map;
}

export function totalButtonCount(project: Project, boardId: string): number {
  return project.controls
    .filter((c) => c.boardId === boardId)
    .reduce((sum, c) => sum + buttonCount(c), 0);
}

File: server/src/codegen/button_index.rs
use std::collections::HashMap;

use crate::model::types::{Control, EncoderMode, Project};

/// Returns (control_id, sub_index) → 0-based joystick button number.
/// Stable: sorted by label then id. Encoders contribute 2 buttons (CW=0, CCW=1).
/// Switches contribute 2 buttons (ON=0, OFF=1). Selectors contribute N buttons (one per position).
pub fn assign_button_indices(
    project: &Project,
    board_id: &str,
) -> HashMap<(String, usize), u32> {
    let mut controls: Vec<&Control> = project
        .controls
        .iter()
        .filter(|c| c.board_id() == board_id)
        .collect();

    controls.sort_by(|a, b| a.label().cmp(b.label()).then(a.id().cmp(b.id())));

    let mut map = HashMap::new();
    let mut next_button: u32 = 0;

    for control in controls {
        let id = control.id().to_string();
        let count = button_count(control);
        for sub in 0..count {
            map.insert((id.clone(), sub), next_button);
            next_button += 1;
        }
    }

    map
}

pub fn total_button_count(project: &Project, board_id: &str) -> u32 {
    project
        .controls
        .iter()
        .filter(|c| c.board_id() == board_id)
        .map(|c| button_count(c) as u32)
        .sum()
}

fn button_count(control: &Control) -> usize {
    match control {
        Control::Button(_) => 1,
        Control::Switch(_) => 2,
        Control::Selector(s) => s.positions.len(),
        Control::Encoder(e) => match e.encoder.mode {
            EncoderMode::Buttons => 2,
            EncoderMode::Axis => 0,
        },
        Control::Analog(_) => 0,
    }
}
